src/find/matchers/logical_matchers.rs

```diff
@@ -0,0 +1,477 @@
+//! This modules contains the matchers used for combining other matchers and
+//! performing boolean logic on them (and a couple of trivial always-true and
+//! always-false matchers). The design is strongly tied to the precedence rules
+//! when parsing command-line options (e.g. "-foo -o -bar -baz" is equivalent
+//! to "-foo -o ( -bar -baz )", not "( -foo -o -bar ) -baz").
+use std::error::Error;
+use std::iter::Iterator;
+use walkdir::DirEntry;
+
+use find::matchers::{Matcher, MatcherIO};
+
+/// This matcher contains a collection of other matchers. A file only matches
+/// if it matches ALL the contained sub-matchers. For sub-matchers that have
+/// side effects, the side effects occur in the same order as the sub-matchers
+/// were pushed into the collection.
+pub struct AndMatcher {
+    submatchers: Vec<Box<Matcher>>,
+}
+
+impl AndMatcher {
+    pub fn new(submatchers: Vec<Box<Matcher>>) -> AndMatcher {
+        AndMatcher { submatchers: submatchers }
+    }
+}
+
+
+impl Matcher for AndMatcher {
+    /// Returns true if all sub-matchers return true. Short-circuiting does take
+    /// place. If the nth sub-matcher returns false, then we immediately return
+    /// and don't make any further calls.
+    fn matches(&self, dir_entry: &DirEntry, matcher_io: &mut MatcherIO) -> bool {
+        self.submatchers.iter().all(|ref x| x.matches(dir_entry, matcher_io))
+    }
+
+    fn has_side_effects(&self) -> bool {
+        self.submatchers.iter().any(|ref x| x.has_side_effects())
+    }
+}
+
+pub struct AndMatcherBuilder {
+    submatchers: Vec<Box<Matcher>>,
+}
+
+impl AndMatcherBuilder {
+    pub fn new() -> AndMatcherBuilder {
+        AndMatcherBuilder { submatchers: Vec::new() }
+    }
+
+    pub fn new_and_condition(&mut self, matcher: Box<Matcher>) {
```

Comment:
This was originally called "and" and OrMatcherBuilder's equivalent was originally called "or". But then ListMatcherBuilder didn't have an obvious matching name... I'd like to keep the 3 methods fairly consistent name-wise but am open to suggestions